File: finance_core/services/snapshots.py
from __future__ import annotations

import sqlite3
from datetime import date
from typing import Any


SNAPSHOT_COLUMNS = (
    "id",
    "as_of_date",
    "bank_total",
    "securities_total",
    "wallet_total",
    "credit_card_unbilled",
    "total_assets",
    "memo",
)


def row_to_dict(row: sqlite3.Row | None) -> dict[str, Any] | None:
    if row is None:
        return None
    return {key: row[key] for key in row.keys()}


def empty_snapshot() -> dict[str, Any]:
    return {
        "id": None,
        "as_of_date": date.today().isoformat(),
        "bank_total": 0,
        "securities_total": 0,
        "wallet_total": 0,
        "credit_card_unbilled": 0,
        "total_assets": 0,
        "memo": None,
    }


def calculate_total(
    bank_total: int,
    securities_total: int,
    wallet_total: int,
    credit_card_unbilled: int,
) -> int:
    return bank_total + securities_total + wallet_total - credit_card_unbilled


def get_latest_snapshot(conn: sqlite3.Connection) -> dict[str, Any]:
    row = conn.execute(
        """
        SELECT id, as_of_date, bank_total, securities_total, wallet_total,
               credit_card_unbilled, total_assets, memo
        FROM asset_snapshots
        ORDER BY as_of_date DESC, id DESC
        LIMIT 1
        """
    ).fetchone()
    return row_to_dict(row) or empty_snapshot()


def insert_snapshot(
    conn: sqlite3.Connection,
    *,
    bank_total: int | None = None,
    securities_total: int | None = None,
    wallet_total: int | None = None,
    credit_card_unbilled: int | None = None,
    memo: str | None = None,
    as_of_date: str | None = None,
) -> dict[str, Any]:
    latest = get_latest_snapshot(conn)
    new_snapshot = latest.copy()

    if bank_total is not None:
        new_snapshot["bank_total"] = bank_total
    if securities_total is not None:
        new_snapshot["securities_total"] = securities_total
    if wallet_total is not None:
        new_snapshot["wallet_total"] = wallet_total
    if credit_card_unbilled is not None:
        new_snapshot["credit_card_unbilled"] = credit_card_unbilled

    new_snapshot["as_of_date"] = as_of_date or date.today().isoformat()
    new_snapshot["memo"] = memo
    new_snapshot["total_assets"] = calculate_total(
        new_snapshot["bank_total"],
        new_snapshot["securities_total"],
        new_snapshot["wallet_total"],
        new_snapshot["credit_card_unbilled"],
    )

    cur = conn.execute(
        """
        INSERT INTO asset_snapshots (
          as_of_date, bank_total, securities_total, wallet_total,
          credit_card_unbilled, total_assets, memo
        )
        VALUES (?, ?, ?, ?, ?, ?, ?)
        """,
        (
            new_snapshot["as_of_date"],
            new_snapshot["bank_total"],
            new_snapshot["securities_total"],
            new_snapshot["wallet_total"],
            new_snapshot["credit_card_unbilled"],
            new_snapshot["total_assets"],
            new_snapshot["memo"],
        ),
    )
    new_snapshot["id"] = cur.lastrowid
    return new_snapshot


def format_snapshot(snapshot: dict[str, Any]) -> str:
    return "\n".join(
        [
            f"総資産:        {snapshot['total_assets']:,}円",
            f"銀行残高:       {snapshot['bank_total']:,}円",
            f"証券評価額:    {snapshot['securities_total']:,}円",
            f"財布残高:          {snapshot['wallet_total']:,}円",
            f"カード利用:      -{snapshot['credit_card_unbilled']:,}円",
        ]
    )

File: cfor_rs/src/services/snapshots.rs
use crate::error::Result;
use crate::models::Snapshot;
use rusqlite::Connection;

pub fn get_latest_snapshot(conn: &Connection) -> Result<Snapshot> {
    let result = conn.query_row(
        "SELECT id, as_of_date, bank_total, securities_total, wallet_total,
                credit_card_unbilled, total_assets, memo
         FROM asset_snapshots
         ORDER BY as_of_date DESC, id DESC
         LIMIT 1",
        [],
        |row| {
            Ok(Snapshot {
                id: Some(row.get(0)?),
                as_of_date: row.get(1)?,
                bank_total: row.get(2)?,
                securities_total: row.get(3)?,
                wallet_total: row.get(4)?,
                credit_card_unbilled: row.get(5)?,
                total_assets: row.get(6)?,
                memo: row.get(7)?,
            })
        },
    );
    match result {
        Ok(s) => Ok(s),
        Err(rusqlite::Error::QueryReturnedNoRows) => Ok(Snapshot::default()),
        Err(e) => Err(e.into()),
    }
}

pub struct SnapshotBuilder {
    pub bank_total: Option<i64>,
    pub securities_total: Option<i64>,
    pub wallet_total: Option<i64>,
    pub credit_card_unbilled: Option<i64>,
    pub memo: Option<String>,
    pub as_of_date: Option<String>,
}

impl Default for SnapshotBuilder {
    fn default() -> Self {
        SnapshotBuilder {
            bank_total: None,
            securities_total: None,
            wallet_total: None,
            credit_card_unbilled: None,
            memo: None,
            as_of_date: None,
        }
    }
}

pub fn insert_snapshot(conn: &Connection, builder: SnapshotBuilder) -> Result<Snapshot> {
    let mut latest = get_latest_snapshot(conn)?;

    if let Some(v) = builder.bank_total {
        latest.bank_total = v;
    }
    if let Some(v) = builder.securities_total {
        latest.securities_total = v;
    }
    if let Some(v) = builder.wallet_total {
        latest.wallet_total = v;
    }
    if let Some(v) = builder.credit_card_unbilled {
        latest.credit_card_unbilled = v;
    }

    latest.as_of_date = builder
        .as_of_date
        .unwrap_or_else(|| chrono::Local::now().format("%Y-%m-%d").to_string());
    latest.memo = builder.memo;
    latest.calculate_total();

    conn.execute(
        "INSERT INTO asset_snapshots
         (as_of_date, bank_total, securities_total, wallet_total,
          credit_card_unbilled, total_assets, memo)
         VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7)",
        rusqlite::params![
            latest.as_of_date,
            latest.bank_total,
            latest.securities_total,
            latest.wallet_total,
            latest.credit_card_unbilled,
            latest.total_assets,
            latest.memo,
        ],
    )?;
    latest.id = Some(conn.last_insert_rowid());
    Ok(latest)
}

pub fn format_snapshot(snapshot: &Snapshot) -> String {
    crate::services::now::format_current_position(snapshot)
}
